### Explore the structure of folders

In [1]:
%%bash
ls *

LICENSE
README.md
data_exploration.ipynb

Synthea_data:
synthea_sample_data_csv_100_latest
synthea_sample_data_csv_1K_nov2021

UCI_ML_data:
diabetes


In [2]:
%%bash
ls UCI_ML_data/

diabetes


In [3]:
%%bash
file -d UCI_ML_data/diabetes/

UCI_ML_data/diabetes/: directory


In [4]:
%%bash
ls -lsah UCI_ML_data/diabetes/

total 32
 0 drwx------@  6 jiz  staff   192B Dec  5 20:08 .
 0 drwxr-xr-x   4 jiz  staff   128B Dec  5 19:14 ..
16 -rw-r--r--@  1 jiz  staff   6.0K Dec  6 01:44 .DS_Store
 0 drwxr-xr-x@ 75 jiz  staff   2.3K Aug 27  1993 Diabetes-Data
 8 -rwx------@  1 jiz  staff   115B May 22  2023 Index
 8 -rwx------@  1 jiz  staff   682B May 22  2023 README


In [5]:
%%bash
ls -lh UCI_ML_data/diabetes/Diabetes-Data/

total 1600
-rw-r--r--@ 1 jiz  staff   1.7K Aug  6  1993 Data-Codes
-rw-r--r--@ 1 jiz  staff   6.1K Aug 27  1993 Domain-Description
-rw-r--r--@ 1 jiz  staff   2.5K Aug 27  1993 README-DIABETES
-rw-r--r--@ 1 jiz  staff    22K Aug  6  1993 data-01
-rw-r--r--@ 1 jiz  staff    18K Aug  6  1993 data-02
-rw-r--r--@ 1 jiz  staff   6.7K Aug  6  1993 data-03
-rw-r--r--@ 1 jiz  staff   6.7K Aug  6  1993 data-04
-rw-r--r--@ 1 jiz  staff   6.7K Aug  6  1993 data-05
-rw-r--r--@ 1 jiz  staff   3.5K Aug  6  1993 data-06
-rw-r--r--@ 1 jiz  staff   5.7K Aug  6  1993 data-07
-rw-r--r--@ 1 jiz  staff   4.0K Aug  6  1993 data-08
-rw-r--r--@ 1 jiz  staff   4.6K Aug  6  1993 data-09
-rw-r--r--@ 1 jiz  staff   5.8K Aug  6  1993 data-10
-rw-r--r--@ 1 jiz  staff   5.2K Aug  6  1993 data-11
-rw-r--r--@ 1 jiz  staff   6.6K Aug  6  1993 data-12
-rw-r--r--@ 1 jiz  staff   6.6K Aug  6  1993 data-13
-rw-r--r--@ 1 jiz  staff   5.1K Aug  6  1993 data-14
-rw-r--r--@ 1 jiz  staff   6.6K Aug  6  1993 data-15
-rw-r--r--@ 1

### Import the necessary libraries to examine the datasets

In [6]:
import pandas as pd
import numpy as np
import pyampute as imputor
from gensim.models import KeyedVectors as GensimKeyedVecs # unsupervised semantic modeling from plain text
# to explore synonyms, hypernyms, and hyponyms
from nltk.corpus import wordnet as NltkWordNet # linguistically-grounded approach finding the semantic relationships between words

#### Example of using Gensim's KeyedVectors
```python
from gensim.models import KeyedVectors
model = KeyedVectors.load_word2vec_format('model_path', binary=True)
similarity = model.similarity('column_name_one', 'column_name_two')
```

#### Example of using NLTK's wordnet
```python
from nltk.corpus import wordnet as wn
word1 = wn.synsets('column_name_one')[0]  # Assuming the first synset
word2 = wn.synsets('column_name_two')[0]
similarity = word1.wup_similarity(word2)  # Wu-Palmer similarity
```

## Dataset exploration

#### UCI diabetes dataset

In [19]:
%%bash
cat UCI_ML_data/diabetes/Diabetes-Data/README-DIABETES

The DIABETES data sets in this directory are provided for use in 1994 
AI in Medicine symposium submissions.  Permission is granted to use the
data sets for other research purposes as long as appropriate credit is
given as to the source (AIM-94 data set provided by Michael Kahn, MD, PhD, 
Washington University, St. Louis, MO).


Index:
------

* Data-Codes: a listing of the codes used in the data sets.

* Domain-Description: This file describes the basic physiology and patho-
physiology of diabetes mellitus and its treatment.

* data-[01-70]: data sets covering several weeks' to months' worth of
outpatient care on 70 patients.  An additional 10 sets will be made
available two weeks prior to the symposium for interested parties.  Please
contact the organizers if you would like to obtain these data sets.


Methods:
--------

You do not need to use all the data in order to participate.  Use any 
subset of the available data from either the ICU data set or the diabetes 
data set.  Furtherm

In [21]:
%%bash
cat UCI_ML_data/diabetes/Diabetes-Data/Data-Codes

Diabetes patient records were obtained from two sources:  an automatic
electronic recording device and paper records.  The automatic device
had an internal clock to timestamp events, whereas the paper records
only provided "logical time" slots (breakfast, lunch, dinner,
bedtime).  For paper records, fixed times were assigned to breakfast
(08:00), lunch (12:00), dinner (18:00), and bedtime (22:00).  Thus
paper records have fictitious uniform recording times whereas
electronic records have more realistic time stamps.

Diabetes files consist of four fields per record.  Each field is
separated by a tab and each record is separated by a newline.

File Names and format:
(1) Date in MM-DD-YYYY format
(2) Time in XX:YY format
(3) Code
(4) Value

The Code field is deciphered as follows:

33 = Regular insulin dose
34 = NPH insulin dose
35 = UltraLente insulin dose
48 = Unspecified blood glucose measurement
57 = Unspecified blood glucose measurement
58 = Pre-breakfast blood glucose measurement
59

In [22]:
# create the column names from the description above
diabetes_columns = ["Date", "Time", "Code", "Value"]

In [23]:
uci_diabetes_df1 = pd.read_csv("UCI_ML_data/diabetes/Diabetes-Data/data-01", sep="\t", names=diabetes_columns)
uci_diabetes_df2 = pd.read_csv("UCI_ML_data/diabetes/Diabetes-Data/data-02", sep="\t", names=diabetes_columns)

In [40]:
uci_diabetes_df1.head()

,Date,Time,Code,Value
0,04-21-1991,9:09,58,100
1,04-21-1991,9:09,33,9
2,04-21-1991,9:09,34,13
3,04-21-1991,17:08,62,119
4,04-21-1991,17:08,33,7


In [42]:
uci_diabetes_df1.tail()

,Date,Time,Code,Value
938,09-02-1991,17:30,33,7
939,09-02-1991,23:00,48,155
940,09-03-1991,7:20,58,110
941,09-03-1991,7:20,33,9
942,09-03-1991,7:20,34,16


In [41]:
uci_diabetes_df2.head()

,Date,Time,Code,Value
0,10-10-1989,08:00,58,149
1,10-10-1989,08:00,33,010
2,10-10-1989,12:00,60,116
3,10-10-1989,12:00,33,004
4,10-10-1989,18:00,62,304


In [72]:
uci_diabetes_df2.head()

,Date,Time,Code,Value
0,10-10-1989,08:00,58,149
1,10-10-1989,08:00,33,10
2,10-10-1989,12:00,60,116
3,10-10-1989,12:00,33,4
4,10-10-1989,18:00,62,304


In [43]:
uci_diabetes_df2.tail()

,Date,Time,Code,Value
756,01-12-1990,22:00,33,016
757,01-13-1990,08:00,58,272
758,01-13-1990,08:00,33,012
759,01-13-1990,12:00,60,388
760,01-13-1990,12:00,33,006


In [55]:
uci_diabetes_df1.dtypes

Date     object
Time     object
Code      int64
Value     int64
dtype: object

In [56]:
uci_diabetes_df2.dtypes

Date     object
Time     object
Code      int64
Value    object
dtype: object

change the datatype of columns in <code>uci_diabetes_df1</code> and <code>uci_diabetes_df2</code> before using <code>pandas.DataFrame.describe()</code>

In [61]:
uci_diabetes_df2["Value"]

0      149
1       10
2      116
3        4
4      304
      ... 
756     16
757    272
758     12
759    388
760      6
Name: Value, Length: 761, dtype: object

In [65]:
# change the datatype of columns in <code>uci_diabetes_df1</code> and <code>uci_diabetes_df2</code> before using DataFrame.describe()
for df in [uci_diabetes_df1, uci_diabetes_df2]:
    df["Code"] = df["Code"].astype("category")

uci_diabetes_df2["Value"] = uci_diabetes_df2["Value"].str.replace(r'^0+', '', regex=True)
uci_diabetes_df2["Value"] = uci_diabetes_df2["Value"].astype("int64") 

ValueError: invalid literal for int() with base 10: 'Hi'

Try to figure out the reason of the bug above

In [75]:
uci_diabetes_df2.tail() # check if the strings are correctly reformatted

,Date,Time,Code,Value
756,01-12-1990,22:00,33,16
757,01-13-1990,08:00,58,272
758,01-13-1990,08:00,33,12
759,01-13-1990,12:00,60,388
760,01-13-1990,12:00,33,6


In [70]:
all_digits = uci_diabetes_df2['Value'].astype(str).str.isdigit().all()
all_digits # check if non-digit values are present in this column

False

In [69]:
uci_diabetes_df1.describe()

,Value
count,943.000000
mean,67.410392
std,85.617268
min,0.000000
25%,7.000000
50%,15.000000
75%,122.000000
max,343.000000


In [68]:
uci_diabetes_df2.describe()

,Date,Time,Code,Value
count,761,761,761,761
unique,96,4,6,225
top,10-27-1989,08:00,33,10
freq,9,193,378,143


<b>Conclusions</b>: <br>
From the data exmamination above, we could see that the UCI diabetes is a time-series data consisting of four columns only. It cannot be seriously considered as electronic health records (EHR). Therefore, it shall be only considered as a potential benchmark when compared to other existing algorithms. To justify our adoption of federated learning for imputation, we shall use the data source of proper EHRs and then create/construct a trainable and testable pandas dataframe. <br>
Therefore, let's put off the experiements on the UCI diabetes dataset for the moment

<hr/>

#### Synthea dataset

In [27]:
%%bash
ls Synthea_data/synthea_sample_data_csv_1K_nov2021/

allergies.csv
careplans.csv
claims.csv
claims_transactions.csv
conditions.csv
devices.csv
encounters.csv
imaging_studies.csv
immunizations.csv
medications.csv
observations.csv
organizations.csv
patients.csv
payer_transitions.csv
payers.csv
procedures.csv
providers.csv
supplies.csv


In [30]:
%%bash
ls Synthea_data/synthea_sample_data_csv_1K_nov2021/ | wc -l # check the number of CSV files/subsheets of the dataset

      18


### First of all, investigate the sheet of patients

In [32]:
synthea_1k_patients_df = pd.read_csv("Synthea_data/synthea_sample_data_csv_1K_nov2021/patients.csv")

In [77]:
synthea_1k_patients_df.dtypes

Id                      object
BIRTHDATE               object
DEATHDATE               object
SSN                     object
DRIVERS                 object
PASSPORT                object
PREFIX                  object
FIRST                   object
LAST                    object
SUFFIX                  object
MAIDEN                  object
MARITAL                 object
RACE                    object
ETHNICITY               object
GENDER                  object
BIRTHPLACE              object
ADDRESS                 object
CITY                    object
STATE                   object
COUNTY                  object
ZIP                    float64
LAT                    float64
LON                    float64
HEALTHCARE_EXPENSES    float64
HEALTHCARE_COVERAGE    float64
dtype: object

In [37]:
synthea_1k_patients_df.describe()

,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE
count,618.000000,1163.000000,1163.000000,1.163000e+03,1.163000e+03
mean,2001.263754,42.255261,-71.336063,1.256248e+06,1.717684e+05
std,443.449141,0.286969,0.618013,1.148747e+06,4.885749e+05
min,1001.000000,41.299942,-73.445165,8.259920e+02,0.000000e+00
25%,1835.000000,42.100121,-71.516412,4.426256e+05,1.293600e+03
50%,2118.500000,42.307061,-71.140665,1.142154e+06,1.747219e+04
75%,2190.000000,42.448367,-71.004105,1.900910e+06,2.051068e+05
max,2861.000000,42.888532,-69.978831,2.311655e+07,1.356228e+07


In [38]:
synthea_1k_patients_df.head()

,Id,BIRTHDATE,DEATHDATE,SSN,DRIVERS,PASSPORT,PREFIX,FIRST,LAST,SUFFIX,...,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE
0,b9c610cd-28a6-4636-ccb6-c7a0d2a4cb85,2019-02-17,NaN,999-65-3251,NaN,NaN,NaN,Damon455,Langosh790,NaN,...,Middleborough Massachusetts US,620 Lynch Tunnel Apt 0,Springfield,Massachusetts,Hampden County,1104.0,42.080389,-72.480431,9039.1645,7964.1255
1,c1f1fcaa-82fd-d5b7-3544-c8f9708b06a8,2005-07-04,NaN,999-49-3323,S99941126,NaN,NaN,Thi53,Wunsch504,NaN,...,Danvers Massachusetts US,972 Tillman Branch Suite 48,Bellingham,Massachusetts,Norfolk County,NaN,42.035213,-71.482519,402723.4150,14064.1350
2,339144f8-50e1-633e-a013-f361391c4cff,1998-05-11,NaN,999-10-8743,S99996708,X75063318X,Mr.,Chi716,Greenfelder433,NaN,...,Athens Athens Prefecture GR,1060 Bernhard Crossroad Suite 15,Boston,Massachusetts,Suffolk County,2131.0,42.292557,-71.061160,571935.8725,787.5375
3,d488232e-bf14-4bed-08c0-a82f34b6a197,2003-01-28,NaN,999-56-6057,S99929424,NaN,Ms.,Phillis443,Walter473,NaN,...,Boston Massachusetts US,677 Ritchie Terrace,Hingham,Massachusetts,Plymouth County,2043.0,42.200491,-70.916076,582557.8030,104782.2070
4,217f95a3-4e10-bd5d-fb67-0cfb5e8ba075,1993-12-23,NaN,999-91-4320,S99991143,X44132498X,Mr.,Jerrold404,Herzog843,NaN,...,Boston Massachusetts US,276 Bernier Branch,Revere,Massachusetts,Suffolk County,NaN,42.381875,-70.999286,475826.8550,18067.0950


In [39]:
synthea_1k_patients_df.tail()

,Id,BIRTHDATE,DEATHDATE,SSN,DRIVERS,PASSPORT,PREFIX,FIRST,LAST,SUFFIX,...,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE
1158,409330fa-7ffd-dbfb-4eba-2349d58a6324,1979-02-28,NaN,999-68-5445,S99991649,X72884711X,Mr.,Coy949,Schaden604,NaN,...,Fitchburg Massachusetts US,434 Hickle Throughway Suite 15,Amherst,Massachusetts,Hampshire County,NaN,42.380084,-72.527376,1.050937e+06,346.5075
1159,cb328021-a854-dc94-e7ae-426580477308,1964-05-31,NaN,999-10-6445,S99986790,X86596484X,Mrs.,Sherry479,Barrows492,NaN,...,Boston Massachusetts US,976 Ortiz Orchard,Stoughton,Massachusetts,Norfolk County,NaN,42.081223,-71.126028,8.031471e+05,553335.4420
1160,41862157-5c14-f706-4a94-d2929be969e7,1967-07-12,NaN,999-63-2407,S99976335,X68999803X,Mr.,Vance413,Jakubowski832,NaN,...,Brockton Massachusetts US,534 Strosin Corner,Gardner,Massachusetts,Worcester County,1440.0,42.604215,-72.013243,1.028321e+06,28607.5050
1161,d53c57a5-4480-2481-32ee-b2844a991c9d,1948-07-28,NaN,999-37-8036,S99939062,X79037907X,Mr.,Cody889,Hilll811,NaN,...,Stoneham Massachusetts US,568 Ryan Stravenue,Mashpee,Massachusetts,Barnstable County,NaN,41.577704,-70.507460,2.638801e+05,155042.0520
1162,cb1b2c74-d1c5-997c-6f8b-20ca9f332eef,1958-11-07,NaN,999-17-1411,S99963285,X2822177X,Mr.,Haywood675,Schroeder447,NaN,...,Peabody Massachusetts US,707 Heathcote Wall Suite 15,Marblehead,Massachusetts,Essex County,NaN,42.466700,-70.903323,1.548721e+06,217.1100


Make a simplified <code>patient_df</code> dataframe to exclude the unnecessary columns

In [78]:
patient_cols = list(synthea_1k_patients_df.columns)
patient_cols

['Id',
 'BIRTHDATE',
 'DEATHDATE',
 'SSN',
 'DRIVERS',
 'PASSPORT',
 'PREFIX',
 'FIRST',
 'LAST',
 'SUFFIX',
 'MAIDEN',
 'MARITAL',
 'RACE',
 'ETHNICITY',
 'GENDER',
 'BIRTHPLACE',
 'ADDRESS',
 'CITY',
 'STATE',
 'COUNTY',
 'ZIP',
 'LAT',
 'LON',
 'HEALTHCARE_EXPENSES',
 'HEALTHCARE_COVERAGE']

To check if Id and SSN are mutually replacable. Therefore, let's just use anonymized IDs instead of privacy-sensitive SSNs

In [190]:
UNIQUE_IDS = pd.unique(synthea_1k_patients_df["Id"]).shape[0]
UNIQUE_SSNS = pd.unique(synthea_1k_patients_df["SSN"]).shape[0]
UNIQUE_IDS == UNIQUE_SSNS

True

<b>Conclusions:</b> Yes, these two attributes are mutually replacable

Let's investigate the varity of genders

In [163]:
synthea_1k_patients_df["GENDER"] = synthea_1k_patients_df["GENDER"].astype("category")

In [178]:
pd.unique(synthea_1k_patients_df["GENDER"])

['M', 'F']
Categories (2, object): ['F', 'M']

In [191]:
pd.unique(synthea_1k_patients_df["Id"].loc[synthea_1k_patients_df["GENDER"] == "M"]).shape[0]

547

In [192]:
pd.unique(synthea_1k_patients_df["Id"].loc[synthea_1k_patients_df["GENDER"] == "F"]).shape[0]

616

<b>Stats</b>: 1163 patients, including 547 Males and 616 Females

Let's investigate the difference between <code>RACE<code> and <code>ETHNICITY<code>, which seem to be ambiguous

In [157]:
pd.unique(synthea_1k_patients_df["RACE"])

array(['white', 'black', 'asian', 'hawaiian', 'other', 'native'],
      dtype=object)

In [194]:
for race in pd.unique(synthea_1k_patients_df["RACE"]):
    num = pd.unique(synthea_1k_patients_df["Id"].loc[synthea_1k_patients_df["RACE"] == race]).shape[0]
    print(f"Race \"{race}\": {num} patients")

Race "white": 963 patients
Race "black": 96 patients
Race "asian": 73 patients
Race "hawaiian": 18 patients
Race "other": 11 patients
Race "native": 2 patients


Show the distribution of the races

In [158]:
pd.unique(synthea_1k_patients_df["ETHNICITY"])

array(['nonhispanic', 'hispanic'], dtype=object)

Show the distribution of the ethnicities

In [195]:
for ethnicity in pd.unique(synthea_1k_patients_df["ETHNICITY"]):
    num = pd.unique(synthea_1k_patients_df["Id"].loc[synthea_1k_patients_df["ETHNICITY"] == ethnicity]).shape[0]
    print(f"Ethnicity \"{ethnicity}\": {num} patients")

Ethnicity "nonhispanic": 1058 patients
Ethnicity "hispanic": 105 patients


Therefore, let's investigate the races of non-Hispianic and Hispanic patients respectively

In [201]:
HISPANIC_RACES = list(pd.unique(synthea_1k_patients_df["RACE"].loc[synthea_1k_patients_df["ETHNICITY"]=="hispanic"]))
set(HISPANIC_RACES)

{'asian', 'black', 'hawaiian', 'other', 'white'}

In [200]:
NON_HISPANIC_RACES = list(pd.unique(synthea_1k_patients_df["RACE"].loc[synthea_1k_patients_df["ETHNICITY"]=="nonhispanic"]))
set(NON_HISPANIC_RACES)

{'asian', 'black', 'hawaiian', 'native', 'other', 'white'}

In [199]:
for race in pd.unique(synthea_1k_patients_df["RACE"]):
    for ethnicity in pd.unique(synthea_1k_patients_df["ETHNICITY"]):
        num = pd.unique(synthea_1k_patients_df["Id"].loc[(synthea_1k_patients_df["RACE"] == race) & (synthea_1k_patients_df["ETHNICITY"] == ethnicity)]).shape[0]
        print(f"{ethnicity} {race}: {num} patients")

nonhispanic white: 879 patients
hispanic white: 84 patients
nonhispanic black: 82 patients
hispanic black: 14 patients
nonhispanic asian: 69 patients
hispanic asian: 4 patients
nonhispanic hawaiian: 17 patients
hispanic hawaiian: 1 patients
nonhispanic other: 9 patients
hispanic other: 2 patients
nonhispanic native: 2 patients
hispanic native: 0 patients


Check the number of alive and deceased patients

let's investigate <code>BIRTHPLACE</code> of the patients

In [206]:
pd.unique(synthea_1k_patients_df.BIRTHPLACE).shape[0] # check the number of unique birthplaces among the 1163 patients

317

In [ ]:
useful_patient_cols = ['Id', 'BIRTHDATE', 'DEATHDATE']

In [209]:
synthea_1k_patients_df.loc[synthea_1k_patients_df["DEATHDATE"].notna()].shape[0]

163

In [210]:
synthea_1k_patients_df.loc[synthea_1k_patients_df["DEATHDATE"].isna()].shape[0]

1000

1000 alive patients, and 163 deceased patients

Create a new column <code></code>

Thoughts: create a new column "LIFESPAN" based on the columns <code>BIRTHDATE, DEATHDATE</code>?

In [ ]:
patient_df = 

Conclusion：the field <code>synthea_1k_patients_df['id']</code> would be useful for concatenation with other dataframes

### Then, investigate the files
```bash
conditions.csv
devices.csv
encounters.csv
medications.csv
observations.csv
```
to decide on the columns to include into the final dataframe for analysis

In [108]:
synthea_1k_devices_df = pd.read_csv("Synthea_data/synthea_sample_data_csv_1K_nov2021/devices.csv")
synthea_1k_devices_df.dtypes

START          object
STOP           object
PATIENT        object
ENCOUNTER      object
CODE            int64
DESCRIPTION    object
UDI            object
dtype: object

In [109]:
synthea_1k_devices_df.shape

(89, 7)

In [110]:
synthea_1k_devices_df.head()

,START,STOP,PATIENT,ENCOUNTER,CODE,DESCRIPTION,UDI
0,2017-09-25T17:26:19Z,NaN,d1822991-eb4f-aafe-6a88-cd5279b4385b,bd2f33cd-4a24-6e32-1cd2-bf075ba9ed77,72506001,Implantable defibrillator device (physical ob...,(01)32224845419831(11)170904(17)420919(10)7079...
1,1987-09-21T02:27:51Z,NaN,04f25f73-04b2-469c-3806-540417a0d61c,807a0ee3-a9b1-30e1-3b19-bc2bddf952e8,72506001,Implantable defibrillator device (physical ob...,(01)84789768157847(11)870830(17)120913(10)3106...
2,2015-07-12T12:58:53Z,NaN,e000e5c8-30f7-7a53-a324-7f3e431ba008,5dfe5aec-2194-02ba-e4db-d16b42d28338,72506001,Implantable defibrillator device (physical ob...,(01)87802513365737(11)150621(17)400705(10)2864...
3,1983-07-18T16:34:31Z,NaN,89464607-f8af-06a3-fe2f-227495c12550,ea2d35e3-1cca-d560-f8ee-21bcfd480a9c,72506001,Implantable defibrillator device (physical ob...,(01)71493530635894(11)830627(17)080711(10)6970...
4,2008-04-04T12:43:40Z,NaN,ff0e4d0e-6181-e36e-d817-64dbcaecb5d0,9ac93727-a73c-5479-6d06-35999c01756d,705643001,Coronary artery stent (physical object),(01)60616412587042(11)080314(17)330329(10)1877...


In [111]:
synthea_1k_devices_df.iloc[0]['CODE']

72506001

From <a>http://purl.bioontology.org/ontology/SNOMEDCT/72506001</a>, we could see that the column <code>synthea_1k_devices_df["CODE"]</code> refers to SNOMED-CT codes

<b>Conclusions</b>: the variety of medical devices applied on the patients is not particularly of our interest in imputation for missing data in EHRs, therefore, we wouldn't like to include this dataframe into our work

In [112]:
synthea_1k_devices_df = None # delete this pointer from the memory

<hr/>

In [114]:
synthea_1k_conditions_df = pd.read_csv("Synthea_data/synthea_sample_data_csv_1K_nov2021/conditions.csv")
synthea_1k_conditions_df.dtypes

START          object
STOP           object
PATIENT        object
ENCOUNTER      object
CODE            int64
DESCRIPTION    object
dtype: object

In [115]:
synthea_1k_conditions_df.shape

(38094, 6)

In [116]:
synthea_1k_conditions_df.head()

,START,STOP,PATIENT,ENCOUNTER,CODE,DESCRIPTION
0,2013-06-24,2013-07-02,c1f1fcaa-82fd-d5b7-3544-c8f9708b06a8,0b2794bd-ec2b-d34f-0610-2523b3b7fcf0,10509002,Acute bronchitis (disorder)
1,2016-02-27,2016-03-14,c1f1fcaa-82fd-d5b7-3544-c8f9708b06a8,a6d818dd-0983-fd1c-eefa-3d2295532c45,283371005,Laceration of forearm
2,2016-08-11,2016-08-22,c1f1fcaa-82fd-d5b7-3544-c8f9708b06a8,36d2e781-4655-0a11-1f70-c69856e02019,444814009,Viral sinusitis (disorder)
3,2016-11-27,2016-12-17,c1f1fcaa-82fd-d5b7-3544-c8f9708b06a8,c8eaaf41-958b-31ab-7de5-568cee8751f3,444814009,Viral sinusitis (disorder)
4,2017-02-22,2017-06-02,c1f1fcaa-82fd-d5b7-3544-c8f9708b06a8,6474f606-5a1b-48c0-bbbf-ad6dcbc24d4e,16114001,Fracture of ankle


In [117]:
synthea_1k_conditions_df.iloc[0]['CODE'] 

10509002

From <a>http://purl.bioontology.org/ontology/SNOMEDCT/10509002</a>, we could see that the column synthea_1k_columns_df["CODE"] refers to SNOMED-CT codes

In [155]:
synthea_1k_patients_df.loc[synthea_1k_patients_df['Id'] == str(synthea_1k_conditions_df.iloc[0]['PATIENT'])]

,Id,BIRTHDATE,DEATHDATE,SSN,DRIVERS,PASSPORT,PREFIX,FIRST,LAST,SUFFIX,...,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE
1,c1f1fcaa-82fd-d5b7-3544-c8f9708b06a8,2005-07-04,NaN,999-49-3323,S99941126,NaN,NaN,Thi53,Wunsch504,NaN,...,Danvers Massachusetts US,972 Tillman Branch Suite 48,Bellingham,Massachusetts,Norfolk County,NaN,42.035213,-71.482519,402723.415,14064.135


Therefore, we can see that the patients in the sheets other than <code>patients.csv</code> are concatenable to <code>patients.csv</code> via the column <code>PATIENT</code> together with the column <code>Id</code> in <code>patients.csv</code>

<b>Conclusions</b>: the variety of clinical conditions of the patients coded by SNOMED-CT is particularly of our interest in imputation for missing data in EHRs, therefore, we wouldn like to include this dataframe into our work

Let's create a Python dictionary of SNOMED-CT code terms describing the clinical conditions, using <code>synthea_1k_conditions_df["CODE"]</code> as the keys and <code>synthea_1k_conditions_df["DESCRIPTION"]</code> as the columns

In [165]:
synthea_1k_conditions_df['CODE'] = synthea_1k_conditions_df['CODE'].astype("category")

In [170]:
synthea_1k_conditions_df.dtypes

START            object
STOP             object
PATIENT          object
ENCOUNTER        object
CODE           category
DESCRIPTION      object
dtype: object

In [171]:
from typing import Dict, Any

In [172]:
CONDITIONS_SNOMED_CT_DICT: Dict[Any, Any] = {}

In [173]:
def add_new_dict_terms(ref_dict: Dict[Any, Any], new_key: str, new_value: Any):
    if new_key not in ref_dict:
        ref_dict[new_key] = new_value

In [174]:
synthea_1k_conditions_df[["CODE", "DESCRIPTION"]].apply(lambda x: add_new_dict_terms(CONDITIONS_SNOMED_CT_DICT, x["CODE"], x["DESCRIPTION"]), 
                                                        axis=1)

0        None
1        None
2        None
3        None
4        None
         ... 
38089    None
38090    None
38091    None
38092    None
38093    None
Length: 38094, dtype: object

In [175]:
len(CONDITIONS_SNOMED_CT_DICT.items())

202

In [177]:
list(CONDITIONS_SNOMED_CT_DICT.items())[:5]

[('10509002', 'Acute bronchitis (disorder)'),
 ('283371005', 'Laceration of forearm'),
 ('444814009', 'Viral sinusitis (disorder)'),
 ('16114001', 'Fracture of ankle'),
 ('44465007', 'Sprain of ankle')]